File: p16/p1644/LowestCommonAncestor.java
package leetcode.p16.p1644;


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) {
        this.val = val;
    }
}

public class LowestCommonAncestor {
    private static TreeNode answer = null;

    public static TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        answer = null;
        helper1(root, p, q);
        return answer;
    }

    public static int helper1(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return 0;
        }
        int left, mid, right;
        if (root.val == p.val || root.val == q.val) {
            mid = 1;
        } else {
            mid = 0;
        }
        left = helper1(root.left, p, q);
        right = helper1(root.right, p, q);
        if (mid + left + right >= 2) {
            answer = root;
        }
        if (mid + left + right >= 1) {
            return 1;
        } else {
            return 0;
        }
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(5);
        root.left.left = new TreeNode(6);
        root.left.right = new TreeNode(2);
        root.left.right.left = new TreeNode(7);
        root.left.right.right = new TreeNode(4);
        root.right = new TreeNode(1);
        root.right.left = new TreeNode(0);
        root.right.right = new TreeNode(8);

        TreeNode p = root.left;
        TreeNode q = root.left.right.right;

        System.out.println(lowestCommonAncestor(root, p, q) == root.left);

        q = new TreeNode(10);
        answer = null;
        System.out.println(lowestCommonAncestor(root, p, q) == null);
    }
}

File: p16/p1644/lowest_common_ancestor.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    """
    DFS iterative
    Dec 29, 2024 03:29
    Time Complexity: O(n)
    Space Complexity: O(H)
    """

    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        return self.dfs(root, p, q)[1]

    def dfs(self, root: TreeNode, p: TreeNode, q: TreeNode) -> (int, TreeNode):
        current = 0
        if root == p or root == q:
            current = 1

        left_count = 0
        if root.left:
            left_count, ancestor = self.dfs(root.left, p, q)
            if ancestor:
                return -1, ancestor

        right_count = 0
        if root.right:
            right_count, ancestor = self.dfs(root.right, p, q)
            if ancestor:
                return -1, ancestor

        if current + left_count + right_count == 2:
            return -1, root

        return current + right_count + left_count, None

class Solution1:
    """
    DFS iterative
    Dec 29, 2024 03:29
    Time Complexity: O(n)
    Space Complexity: O(H)
    """

    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        return self.dfs(root, p, q)[1]

    def dfs(self, root: TreeNode, p: TreeNode, q: TreeNode) -> (int, TreeNode):
        current = 0
        if root == p or root == q:
            current = 1

        left_count = 0
        if root.left:
            left_count, ancestor = self.dfs(root.left, p, q)
            if ancestor:
                return -1, ancestor

        right_count = 0
        if root.right:
            right_count, ancestor = self.dfs(root.right, p, q)
            if ancestor:
                return -1, ancestor

        if current + left_count + right_count == 2:
            return -1, root

        return current + right_count + left_count, None
